### ISSUE TITLE
security: allow SessionLog object to be passed to session_log arg

### ISSUE THREAD

**Issue by dannywade** [ktbyers/netmiko#3307]
TL;DR: allow SessionLog object to be passed to session_log argument for BaseConnection

Netmiko logs all channel communication to a file via its own SessionLog for each device interaction. While SessionLog filters the username and password out automatically, there may be a case where you want more words/phrases sanitized in the session log. Currently, you can only pass a string or BufferedIOBase subclass to the session_log argument, which doesn't allow you to edit the SessionLog no_log argument. The no_log argument allows you to pass in a dict with words that you want sanitized from the session log.

By adding the ability to pass in a SessionLog object, a user can customize the no_log argument, which allows for an arbitrary number of words to be sanitized in the logs, not just the username and password.

**Comment by ktbyers** [comment]
Should be done.

**Comment by dannywade** [comment]
@ktbyers - I cloned the repo and tried passing in a SessionLog object and found a weird bug.

Before I get into the bug, I found that self.session_log does need manually opened in order for it to be saved to a file. Below is a snippet of code I had to add in order for the session log to save to a file.

```
            elif isinstance(session_log, SessionLog):
                # SessionLog object
                self.session_log = session_log
                self.session_log.open() # <-- had to add
            else:
                raise ValueError(
                    "session_log must be a path to a file, a file handle, "
                    "SessionLog object, or a BufferedIOBase subclass."
                )
```

Once I added self.session_log.open(), I noticed that the config and the channel log was writing to the file, causing duplicate logging. Here's a snippet of the logs:

```
terminal width 511
danslab-sw01#
danslab-sw01#terminal width 511
danslab-sw01#terminal length 0
terminal length 0
danslab-sw01#

danslab-sw01#

danslab-sw01#configure terminal
configure terminal
Enter configuration commands, one per line.  End with CNTL/Z.
danslab-sw01(config)#

danslab-sw01(config)#logging host ********
logging host 10.1.1.1
danslab-sw01(config)#logging host 10.1.1.2
logging host 10.1.1.2
danslab-sw01(config)#

danslab-sw01(config)#end
end
danslab-sw01#

danslab-sw01#

danslab-sw01#exit
```

And here's the associated script I used for testing:

```
from netmiko import ConnectHandler
from netmiko.session_log import SessionLog


sanitize_secrets = {
    "log_server": "10.1.1.1",
}

nm_session_log = SessionLog(
    file_name=f"device_logs/testing.log",
    no_log=sanitize_secrets,
    record_writes=True,
)

device = {
    "device_type": "cisco_ios",
    "host": "10.1.1.1",
    "username": "cisco",
    "password": "cisco",
    "session_log": nm_session_log,
}

cfg_file = "test_config.txt"
with ConnectHandler(**device) as net_connect:
    output = net_connect.send_config_from_file(cfg_file)
```

I figure one of the Cisco subclasses of BaseConnection is writing the additional config lines in the session log, but wanted your opinion on it before I dive into the code and create another PR.

**Comment by ktbyers** [comment]
You are initializing the SessionLog object with `record_writes=True`. This implies you will see the "written" data and the echoed data. Consequently, you will have a duplication of many things (i.e. these are the duplicates you see in the log file).

Yeah, yes an already open() session log object would need passed in. So the expectation is your code would do the following:

```
nm_session_log = SessionLog(
    file_name=f"device_logs/testing.log",
    no_log=sanitize_secrets,
    record_writes=False,        # changed to False which is also the default
)
nm_session_log.open()
```

I am open to either way...as calling open() on an already open() SessionLog object I believe is a no-op (so it would also probably be fine if we included the `open()` call in the Netmiko code (as you proposed above).

**Comment by dannywade** [comment]
@ktbyers Thanks for responding. I'll move forward with creating a PR to add the `open()` call.

Is there any way to remove the duplication in the session log? Possibly remove the echoed data? I understand what you're saying with the `record_writes=True` argument, but I'm only interested in seeing what's configured on the device, so basically what's "written" to the channel.

**Comment by ktbyers** [comment]
So you don't want to see the actual output from the device (don't want to see prompts, show command output, error messages), you only want to see what you actually write to the device?

Typically what people want is what it would look like if you had "captured the session" (this is usually just the reads i.e. you don't include any writes). Your writes will get echoed anyways so you will see them in the session (with a few exception like passwords which sometimes get masked by the device on echo).

At the end of the day though...now that you can pass in a SessionLog object, you can make your own SessionLog object that behaves the way you want (as long as it conforms to the Netmiko interface i.e. has the methods that Netmiko expects to be supported).

**Comment by dannywade** [comment]
@ktbyers No, you're correct. I only want to see the actual output from the device (prompts, command output, error messages, etc.). Sorry, I didn't explain it right. I guess I'm looking to remove the echoed commands, so you don't see the echoed commands with the actual output.

I added an example with comments of the lines I'd like to be removed from the session log. Besides being duplicate, the output doesn't sanitize the words/phrases outlined in the `no_log` argument (i.e. 10.1.1.1).

```
danslab-sw01(config)#logging host ********
logging host 10.1.1.1 <--- remove
danslab-sw01(config)#logging host 10.1.1.2
logging host 10.1.1.2 <--- remove
danslab-sw01(config)#

danslab-sw01(config)#end
end <--- remove
```

**Comment by ktbyers** [comment]
Yeah, just set `record_writes=False` that should probably give you what you want.

**Comment by dannywade** [comment]
@ktbyers I tested again with `record_writes=False` and that removed the duplication. However, now the "sensitive" words defined in the `no_log` dict are not sanitized from the log. See below.

Referring back to my previous example, the 10.1.1.1 IP should be sanitized in the log.

```
danslab-sw01#
danslab-sw01#terminal width 511
danslab-sw01#terminal length 0
danslab-sw01#
danslab-sw01#
danslab-sw01#configure terminal
Enter configuration commands, one per line.  End with CNTL/Z.
danslab-sw01(config)#
danslab-sw01(config)#logging host 10.1.1.1
danslab-sw01(config)#logging host 10.1.1.2
danslab-sw01(config)#
danslab-sw01(config)#end
danslab-sw01#
danslab-sw01#
danslab-sw01#exit
```

**Comment by ktbyers** [comment]
Okay, let me look at that (that probably is a bug).

**Comment by dannywade** [comment]
@ktbyers ah okay, that makes sense. So I wonder if the password and/or secret (default strings being sanitizied by session log) were long enough to get split up amongst two read channels that they wouldn't get caught by the filter?

**Comment by ktbyers** [comment]
Yeah, this issue would likely happen with secrets and passwords (in cases where the device echoed them to the screen).

So I will need to flag it as a security issue in Netmiko.
